### PR TITLE
Test fastANI matrixes - Issue 104

### LINKED ISSUE #104: Compare fastANI output to reference matrices from legacy pyANI
We already have in place fixtures and tests to run fastANI, but don't yet confirm the values at the end match what pyANI v0.3 was producing.

Cross reference #100 which adds the expected values for ANIm as TSV files.

### PR DESCRIPTION
**Pull request by peterjc**
Would close #104.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

File: pyani_plus/private_cli.py [hunk 1]
```diff
@@ -348,28 +348,27 @@ def log_fastani(  # noqa: PLR0913
...
+    estimated_aln_length = estimated_cov_query * min(query.length, subject.length)
 
     db_orm.add_comparison(
         session,
         configuration_id=config.configuration_id,
         query_hash=query_md5,
         subject_hash=subject_md5,
         identity=identity,
-        aln_length=orthologous_matches,
+        aln_length=estimated_aln_length,
```
Comment (peterjc) [inline pyani_plus/private_cli.py:368]: We don't have a real alignment length from fastANI.

One might argue by analogy with the current treatment of sim-errors (recording the number of left over fragments) that we record the number of orthologous matches?

Or, the estimate in bp using the estimated coverage fraction?

**Comment by peterjc** [comment]
My inclination here is to write a standalone script which parses the 9 sets of fastANI output into the various expected 3x3 matrices which we can check in as fixtures. i.e. Don't use pyANI v0.3 at all. It wouldn't be an independent implementation (if both written by me), but better than nothing as this script can then be used in future should we ever change the example sets.

**Comment by peterjc** [comment]
The PR now tests the fastANI matrices produced via snakemake against fixtures generated by a standalone Python script *without* calling the unfinished method in pyANI v0.3 at all.

Currently it records an estimated alignment length. We could record None here, but that would mean a minor change to the DB ORM/schema. We could also report the number of orthologous matched fragments as the alignment length (not in base pairs like other methods).

**Comment by peterjc** [comment]
I thought I found a way to silence Codacy's complaints about asserts in the test (adding a ``.bandit`` configuration file), and perhaps CodeFactor too. However, this failed. Even locally I could only get bandit 1.7.10 to read from an explicitly configuration file given at the command line (a possible regression).

For the record, ``bandit -c pyproject.toml ...`` worked nicely with this:

```
[tool.bandit]
# B324:hashlib - used for MD5 checksum
# B404:import_subprocess - unavoidable
skips = ["B324", "B404"]

[tool.bandit.assert_used]
# Skip B101 aka assert_used in our test files:
skips = ["*tests/*.py"]
```

In practice I think we get all the tests of interest via ruff anyway - see https://docs.astral.sh/ruff/rules/#flake8-bandit-s

**Comment by peterjc** [comment]
I wasn't expecting a conflict, but have pulled out the ruff check change to #113 instead. I will rebase this once that's merged.

**Comment by peterjc** [comment]
Rebased after #113 was merged, and updated to drop the ``# noqa: T201`` bits on the print lines.

**Comment by peterjc** [comment]
I'm going to rebase this for #114 (merged) and #116 (pending)...

File: pyani_plus/private_cli.py [hunk 2]
```diff
@@ -350,28 +318,30 @@ def log_fastani(  # noqa: PLR0913
...
+
+    # Need to lookup query/subject length to estimate alignment length.
+    # Should not be needed in standard workflow, but also ensures FASTA are in DB:
+    query = db_orm.add_genome(session, query_fasta, query_md5)
     subject = db_orm.add_genome(session, subject_fasta, subject_md5)
-    cov_subject = float(orthologous_matches) / subject.length
+
+    estimated_cov_query = float(orthologous_matches) / fragments  # an approximation
+    sim_errors = fragments - orthologous_matches  # proxy value, not bp
+    estimated_aln_length = int(
```
Comment (widdowquinn) [inline pyani_plus/private_cli.py:331]: The fragment size used by fastANI is a controllable parameter (default: 3000bp) and reported by fastANI. We could use this directly to report alignment length as total length of orthologous fragments, as `fragment_length` * `orthologous_matches`

Comment (peterjc) [reply pyani_plus/private_cli.py:331]: I like that calculation better than the current estimate 😀

Comment (peterjc) [reply pyani_plus/private_cli.py:331]: For query coverage, we currently have used number of orthologous matched fragments divided by total fragments.

Leighton suggests we _could_ use fragment size (default 3000bp) times number of orthologous fragments, divided by query length. Perhaps an option for later.

**Comment by peterjc** [comment]
Agreed to merge this after changing the ``estimated_aln_length`` as discussed above.

**Pull request merged**